In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime
from tqdm import tqdm
from collections import Counter
%matplotlib inline

In [2]:
user_engagement_df = pd.read_csv('relax_challenge/takehome_user_engagement.csv')
user_engagement_df.head()

time_stamp  user_id  visited
0  2014-04-22 03:53:30        1        1
1  2013-11-15 03:45:04        2        1
2  2013-11-29 03:45:04        2        1
3  2013-12-09 03:45:04        2        1
4  2013-12-25 03:45:04        2        1

In [3]:
user_engagement_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207917 entries, 0 to 207916
Data columns (total 3 columns):
time_stamp    207917 non-null object
user_id       207917 non-null int64
visited       207917 non-null int64
dtypes: int64(2), object(1)
memory usage: 4.8+ MB


In [4]:
user_engagement_df['time_stamp'] = pd.to_datetime(user_engagement_df.time_stamp)

In [5]:
user_engagement_df.head()

time_stamp  user_id  visited
0 2014-04-22 03:53:30        1        1
1 2013-11-15 03:45:04        2        1
2 2013-11-29 03:45:04        2        1
3 2013-12-09 03:45:04        2        1
4 2013-12-25 03:45:04        2        1

In [6]:
users_df = pd.read_csv('relax_challenge/takehome_users.csv')
users_df.head()

object_id        creation_time               name  \
0          1  2014-04-22 03:53:30     Clausen August   
1          2  2013-11-15 03:45:04      Poole Matthew   
2          3  2013-03-19 23:14:52  Bottrill Mitchell   
3          4  2013-05-21 08:09:28    Clausen Nicklas   
4          5  2013-01-17 10:14:20          Raw Grace   

                        email creation_source  last_session_creation_time  \
0    AugustCClausen@yahoo.com    GUEST_INVITE                1.398139e+09   
1      MatthewPoole@gustr.com      ORG_INVITE                1.396238e+09   
2  MitchellBottrill@gustr.com      ORG_INVITE                1.363735e+09   
3   NicklasSClausen@yahoo.com    GUEST_INVITE                1.369210e+09   
4          GraceRaw@yahoo.com    GUEST_INVITE                1.358850e+09   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         1                           0      11   
1                         0                           0       1   
2                         0                           0      94   
3                         0                           0       1   
4                         0                           0     193   

   invited_by_user_id  
0             10803.0  
1               316.0  
2              1525.0  
3              5151.0  
4              5240.0

In [7]:
users_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 10 columns):
object_id                     12000 non-null int64
creation_time                 12000 non-null object
name                          12000 non-null object
email                         12000 non-null object
creation_source               12000 non-null object
last_session_creation_time    8823 non-null float64
opted_in_to_mailing_list      12000 non-null int64
enabled_for_marketing_drip    12000 non-null int64
org_id                        12000 non-null int64
invited_by_user_id            6417 non-null float64
dtypes: float64(2), int64(4), object(4)
memory usage: 937.6+ KB


### Users who have used atleast once

In [8]:
user_engagement_df.user_id.nunique()

8823

### Total number of users signed up for the application

In [9]:
users_df.object_id.nunique()

12000

### Users who have signed up but not used even once. These are inactive users

In [10]:
users_df.object_id.nunique() - user_engagement_df.user_id.nunique()

3177

In [11]:
users_df['last_session_creation_time'].isnull().value_counts()

False    8823
True     3177
Name: last_session_creation_time, dtype: int64

## Missing Data

In [12]:
users_df.isnull().any()

object_id                     False
creation_time                 False
name                          False
email                         False
creation_source               False
last_session_creation_time     True
opted_in_to_mailing_list      False
enabled_for_marketing_drip    False
org_id                        False
invited_by_user_id             True
dtype: bool

In [13]:
users_df['invited_by_user_id'].isnull().value_counts()

False    6417
True     5583
Name: invited_by_user_id, dtype: int64

There are 5583 users who have not been invited by any of the existing users.

In [14]:
user_engagement_df.isnull().any()

time_stamp    False
user_id       False
visited       False
dtype: bool

There are no missing data in user_engagement_df

In [15]:
user_engagement_df['visited'][0]

1

Need to identify an adopted user based on whether they have logged in atleast thrice in a week. To do that, the easy way is to make the dataframe to be indexed on timestamp column and then use grouper from pandas to groupby frequency of a week/7days.

In [16]:
#convert to time stamp index
user_engagement_df.index=user_engagement_df.time_stamp
user_engagement_df.drop(labels='time_stamp',axis=1,inplace=True)

In [17]:
user_engagement_df.head()

user_id  visited
time_stamp                           
2014-04-22 03:53:30        1        1
2013-11-15 03:45:04        2        1
2013-11-29 03:45:04        2        1
2013-12-09 03:45:04        2        1
2013-12-25 03:45:04        2        1

In [18]:
#Group by user_id and resample to 1 week period, sum over period
#https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Grouper.html
df_agg = user_engagement_df.groupby([pd.Grouper(freq='W'),'user_id']).sum()

In [29]:
df_agg.head()

visited
time_stamp user_id         
2012-06-03 563            1
           1693           1
           1995           1
           2120           1
           2136           1

In [30]:
df_agg.tail()

visited
time_stamp user_id         
2014-06-08 11869          1
           11885          1
           11895          3
           11906          1
           11924          3

In [31]:
#find all user id's with a sum of 3 or more indicating an adopted user
#to remove one level of grouping
df_adopt = df_agg[df_agg.visited>=3].unstack(level=1).melt()

df_adopt.head()

NaN  user_id  value
0  visited     1693    3.0
1  visited     1693    NaN
2  visited     1693    3.0
3  visited     1693    NaN
4  visited     1693    3.0

In [20]:
adopted_users = pd.DataFrame(df_adopt.user_id.unique(),index=range(df_adopt.user_id.unique().shape[0]),columns=['user_id'])
len(adopted_users)

1445

In [32]:
adopted_users.head()

user_id
0     1693
1      728
2    11764
3     5297
4     6171

 There are 1445 adopted users

In [21]:
#create df of features for adopted users only
df_join = users_df.merge(adopted_users,how='inner',left_on='object_id',right_on='user_id')
df_join.head()

object_id        creation_time            name  \
0          2  2013-11-15 03:45:04   Poole Matthew   
1         10  2013-01-16 22:08:03    Santos Carla   
2         20  2014-03-06 11:46:38   Helms Mikayla   
3         33  2014-03-11 06:29:09     Araujo José   
4         42  2012-11-11 19:05:07  Pinto Giovanna   

                           email creation_source  last_session_creation_time  \
0         MatthewPoole@gustr.com      ORG_INVITE                1.396238e+09   
1  CarlaFerreiraSantos@gustr.com      ORG_INVITE                1.401833e+09   
2             lqyvjilf@uhzdq.com          SIGNUP                1.401364e+09   
3     JoseMartinsAraujo@cuvox.de    GUEST_INVITE                1.401518e+09   
4    GiovannaCunhaPinto@cuvox.de          SIGNUP                1.401045e+09   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         0                           0       1   
1                         1                           1     318   
2                         0                           0      58   
3                         0                           0     401   
4                         1                           0     235   

   invited_by_user_id  user_id  
0               316.0        2  
1              4143.0       10  
2                 NaN       20  
3                79.0       33  
4                 NaN       42

In [22]:
#drop irrelevant columns like objectid, creation timename and email, userid
drop_cols = list(df_join.columns[0:4])
drop_cols.append('user_id')
df_join = df_join.drop(drop_cols,axis=1)



In [23]:
#fill na values in invited_by_user column
df_join['invited_by_user_id'].fillna(value=0,inplace=True)

In [34]:
df_join.head()

creation_source  last_session_creation_time  opted_in_to_mailing_list  \
0      ORG_INVITE                1.396238e+09                         0   
1      ORG_INVITE                1.401833e+09                         1   
2          SIGNUP                1.401364e+09                         0   
3    GUEST_INVITE                1.401518e+09                         0   
4          SIGNUP                1.401045e+09                         1   

   enabled_for_marketing_drip  org_id  invited_by_user_id  
0                           0       1               316.0  
1                           1     318              4143.0  
2                           0      58                 0.0  
3                           0     401                79.0  
4                           0     235                 0.0

Assuming creations source as an important feature we crate a dummy dataframe from creation_source column using get_dummies function in pandas to create a dataframe with all the different variable available in the creation_source column. This will then be appended to the main data frame so that each variable can be used as a feature

In [33]:
#https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.get_dummies.html

df_create = pd.get_dummies(df_join['creation_source'])
df_create.head()

GUEST_INVITE  ORG_INVITE  PERSONAL_PROJECTS  SIGNUP  SIGNUP_GOOGLE_AUTH
0             0           1                  0       0                   0
1             0           1                  0       0                   0
2             0           0                  0       1                   0
3             1           0                  0       0                   0
4             0           0                  0       1                   0

In [35]:
#one hot encode creation_source feature

df_features = pd.concat([df_join,df_create],axis=1)
df_features.drop('creation_source',axis=1,inplace=True)

#convert columns to float64
for col in df_features.columns:
    df_features[col] = df_features[col].astype('float64')
df_features.head()

last_session_creation_time  opted_in_to_mailing_list  \
0                1.396238e+09                       0.0   
1                1.401833e+09                       1.0   
2                1.401364e+09                       0.0   
3                1.401518e+09                       0.0   
4                1.401045e+09                       1.0   

   enabled_for_marketing_drip  org_id  invited_by_user_id  GUEST_INVITE  \
0                         0.0     1.0               316.0           0.0   
1                         1.0   318.0              4143.0           0.0   
2                         0.0    58.0                 0.0           0.0   
3                         0.0   401.0                79.0           1.0   
4                         0.0   235.0                 0.0           0.0   

   ORG_INVITE  PERSONAL_PROJECTS  SIGNUP  SIGNUP_GOOGLE_AUTH  
0         1.0                0.0     0.0                 0.0  
1         1.0                0.0     0.0                 0.0  
2         0.0                0.0     1.0                 0.0  
3         0.0                0.0     0.0                 0.0  
4         0.0                0.0     1.0                 0.0

### Principle Component Analysis

Then we perform a standard Princple Component Analysis (PCA) to find the most influential feature.Principal component analysis (PCA) is a technique used to emphasize variation and bring out strong patterns in a dataset. It's often used to make data easy to explore and visualize.

In [25]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

#scale data
scaler = StandardScaler()
features = scaler.fit_transform(df_features)

#fit PCA
pca = PCA()
components = pca.fit_transform(features)

In [26]:
np.sum(pca.explained_variance_ratio_[0:7])

0.92278033675256

In [27]:
df_comp = pd.DataFrame(pca.components_,columns=df_features.columns,index=['PC-1','PC-2','PC-3','PC-4','PC-5','PC-6','PC-7','PC-8','PC-9','PC-10'])
#absolute values of correlation with principal components
best_features = np.absolute(df_comp[np.absolute(df_comp) > 0.1])
best_features.head()

last_session_creation_time  opted_in_to_mailing_list  \
PC-1                         NaN                       NaN   
PC-2                         NaN                  0.695712   
PC-3                    0.116892                       NaN   
PC-4                         NaN                       NaN   
PC-5                    0.230407                       NaN   

      enabled_for_marketing_drip    org_id  invited_by_user_id  GUEST_INVITE  \
PC-1                         NaN       NaN            0.637653      0.232628   
PC-2                     0.69891       NaN                 NaN           NaN   
PC-3                         NaN       NaN                 NaN      0.784060   
PC-4                         NaN  0.261476                 NaN           NaN   
PC-5                         NaN  0.219055                 NaN           NaN   

      ORG_INVITE  PERSONAL_PROJECTS    SIGNUP  SIGNUP_GOOGLE_AUTH  
PC-1    0.486575           0.233976  0.390324            0.296733  
PC-2         NaN                NaN       NaN                 NaN  
PC-3    0.590037                NaN       NaN                 NaN  
PC-4         NaN           0.108246  0.648711            0.697782  
PC-5         NaN           0.825688  0.296054            0.350811

In [28]:
#sum correlations to get a relative estimate of the feature importance
best_features.head(7).sum(axis=0).sort_values(ascending=False)

org_id                        1.792770
SIGNUP_GOOGLE_AUTH            1.705164
last_session_creation_time    1.700941
SIGNUP                        1.524752
PERSONAL_PROJECTS             1.438816
ORG_INVITE                    1.076612
GUEST_INVITE                  1.016688
enabled_for_marketing_drip    0.698910
opted_in_to_mailing_list      0.695712
invited_by_user_id            0.637653
dtype: float64